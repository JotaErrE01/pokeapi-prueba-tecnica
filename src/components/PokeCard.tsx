import { FC } from "react";
import { LazyLoadImage } from 'react-lazy-load-image-component';
import { Link } from "react-router-dom";
import { PokeResults } from "../interfaces/IPokeList";

interface Props {
  poke: PokeResults;
  index: number;
}

export const PokeCard: FC<Props> = ({ poke, index }) => {
  return (
    <Link
      to={`/${poke.name}`}
      className="card"
    >
      <LazyLoadImage
        src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${index + 1}.svg`}
        className="card__image"
        alt={poke.name}
        loading="lazy"
      />
      <h2
        className="card__title"
      >{poke.name}</h2>
    </Link>
  )
};
