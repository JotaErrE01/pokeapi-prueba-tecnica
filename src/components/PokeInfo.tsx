import { useNavigate, useParams } from 'react-router';
import { usePokeApi } from '../hooks/usePokeApi';
import { IPokeInfo } from '../interfaces/IPokeInfo';

export const PokeInfo = () => {
  const { name } = useParams();
  const { error, isLoading, data } = usePokeApi<IPokeInfo>(`/pokemon/${name}`);

  if (isLoading) return <></>;

  if (error) {
    return <h1
      className='title'
    >Lo sentimos, Ocurri√≥ un errror inesperado</h1>;
  };

  return (
    <div>
      <h1
        className='title'
      >{data?.name}</h1>

      <div
        className='container card__info'
      >
        <img
          className='main_card__image'
          src={data?.sprites.other?.dream_world.front_default}
          alt={data?.name}
        />

        <div className='text_info'>
          <h2
            className='subtitle'
          >Abilities</h2>
          <ul className='abilities' >
            {
              data?.abilities.map((ability, index) => (
                <li
                  key={index}
                >{ability.ability.name}</li>
              ))
            }
          </ul>
        </div>

        <div>
          <h2
            className='subtitle'
          >Sprites</h2>
          <div
            className='card__info__images__container'
          >
            <img
              className='card__info__image'
              src={data?.sprites.front_default}
              alt={data?.name}
            />

            <img
              className='card__info__image'
              src={data?.sprites.back_default}
              alt={data?.name}
            />
            
            <img
              className='card__info__image'
              src={data?.sprites.front_shiny}
              alt={data?.name}
            />

            <img
              className='card__info__image'
              src={data?.sprites.back_shiny}
              alt={data?.name}
            />
          </div>
        </div>
      </div>
    </div>
  )
};
