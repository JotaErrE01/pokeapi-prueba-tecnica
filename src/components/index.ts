export * from './PokeInfo';
export * from './PokeList';
