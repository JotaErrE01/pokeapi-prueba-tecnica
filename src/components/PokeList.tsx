import { useRef, useState, useEffect } from 'react';
import { usePokeApi } from "../hooks";
import { IPokeList } from "../interfaces/IPokeList";
import { PokeCard } from "./PokeCard";
import { Spinner } from './Spinner';

export const PokeList = () => {
  const [limit, setLimit] = useState(21);
  const ref = useRef<HTMLDivElement>(null);
  const {
    error,
    isLoading,
    data,
  } = usePokeApi<IPokeList>(`/pokemon?limit=${limit}`);

  useEffect(() => {
    window.addEventListener('scroll', () => {
      if (window.scrollY + window.innerHeight >= ref.current!.offsetHeight) {
        setLimit(state => state + 21);
      }
    });

    return () => {
      window.removeEventListener('scroll', () => {
        if (window.scrollY + window.innerHeight >= ref.current!.offsetHeight) {
          setLimit(21);
        }
      });
    }
  }, [ref]);

  if (error) {
    return (
      <div
        className="error"
      >
        <h1>Lo sentimos, Intente de Nuevo.</h1>
      </div>
    )
  }

  return (
    <>
      <h1 className="title"
      >Pokedex</h1>
      <div
        id="pokeList"
        className="container cardList"
        ref={ref}
      >
        {
          isLoading && !data ? (
            <Spinner />
          )
            :
            data?.results.map((poke, index) => (
              <PokeCard
                key={poke.name}
                poke={poke}
                index={index}
              />
            ))
        }
      </div>

      {
        isLoading && (
          <Spinner />
        )
      }
    </>
  )
};
