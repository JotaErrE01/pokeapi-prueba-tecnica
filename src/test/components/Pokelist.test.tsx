import { test, describe, expect } from 'vitest';
import { PokeList } from '../../components/PokeList';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

describe('Pruebas en el componente Pokelist', () => {
  test('Debe de mostrar el spinner al cargar', () => {
    const { container } = render(<PokeList />, { wrapper: BrowserRouter });
    expect(container).toMatchSnapshot();
  });

  test('Debe hacer match con el sanpshot de pokemons y tener 21', async () => {
    const { container } = render(<PokeList />, { wrapper: BrowserRouter });
    await new Promise((resolve) => setTimeout(resolve, 200));

    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('.card').length).toBe(21);
  });
});
