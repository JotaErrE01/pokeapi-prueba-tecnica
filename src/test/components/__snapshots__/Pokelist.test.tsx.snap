// Vitest Snapshot v1

exports[`Pruebas en el componente Pokelist > Debe de mostrar el spinner al cargar 1`] = `
<div>
  <h1
    class="title"
  >
    Pokedex
  </h1>
  <div
    class="container cardList"
    id="pokeList"
  >
    <div
      class="loading-container"
    >
      <div
        class="sk-chase"
      >
        <div
          class="sk-chase-dot"
        />
        <div
          class="sk-chase-dot"
        />
        <div
          class="sk-chase-dot"
        />
        <div
          class="sk-chase-dot"
        />
        <div
          class="sk-chase-dot"
        />
        <div
          class="sk-chase-dot"
        />
      </div>
    </div>
  </div>
  <div
    class="loading-container"
  >
    <div
      class="sk-chase"
    >
      <div
        class="sk-chase-dot"
      />
      <div
        class="sk-chase-dot"
      />
      <div
        class="sk-chase-dot"
      />
      <div
        class="sk-chase-dot"
      />
      <div
        class="sk-chase-dot"
      />
      <div
        class="sk-chase-dot"
      />
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente Pokelist > Debe hacer match con el sanpshot de pokemons y tener 21 1`] = `
<div>
  <h1
    class="title"
  >
    Pokedex
  </h1>
  <div
    class="container cardList"
    id="pokeList"
  >
    <a
      class="card"
      href="/bulbasaur"
    >
      <img
        alt="bulbasaur"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
      />
      <h2
        class="card__title"
      >
        bulbasaur
      </h2>
    </a>
    <a
      class="card"
      href="/ivysaur"
    >
      <img
        alt="ivysaur"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/2.svg"
      />
      <h2
        class="card__title"
      >
        ivysaur
      </h2>
    </a>
    <a
      class="card"
      href="/venusaur"
    >
      <img
        alt="venusaur"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/3.svg"
      />
      <h2
        class="card__title"
      >
        venusaur
      </h2>
    </a>
    <a
      class="card"
      href="/charmander"
    >
      <img
        alt="charmander"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/4.svg"
      />
      <h2
        class="card__title"
      >
        charmander
      </h2>
    </a>
    <a
      class="card"
      href="/charmeleon"
    >
      <img
        alt="charmeleon"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/5.svg"
      />
      <h2
        class="card__title"
      >
        charmeleon
      </h2>
    </a>
    <a
      class="card"
      href="/charizard"
    >
      <img
        alt="charizard"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/6.svg"
      />
      <h2
        class="card__title"
      >
        charizard
      </h2>
    </a>
    <a
      class="card"
      href="/squirtle"
    >
      <img
        alt="squirtle"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/7.svg"
      />
      <h2
        class="card__title"
      >
        squirtle
      </h2>
    </a>
    <a
      class="card"
      href="/wartortle"
    >
      <img
        alt="wartortle"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/8.svg"
      />
      <h2
        class="card__title"
      >
        wartortle
      </h2>
    </a>
    <a
      class="card"
      href="/blastoise"
    >
      <img
        alt="blastoise"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/9.svg"
      />
      <h2
        class="card__title"
      >
        blastoise
      </h2>
    </a>
    <a
      class="card"
      href="/caterpie"
    >
      <img
        alt="caterpie"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/10.svg"
      />
      <h2
        class="card__title"
      >
        caterpie
      </h2>
    </a>
    <a
      class="card"
      href="/metapod"
    >
      <img
        alt="metapod"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/11.svg"
      />
      <h2
        class="card__title"
      >
        metapod
      </h2>
    </a>
    <a
      class="card"
      href="/butterfree"
    >
      <img
        alt="butterfree"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/12.svg"
      />
      <h2
        class="card__title"
      >
        butterfree
      </h2>
    </a>
    <a
      class="card"
      href="/weedle"
    >
      <img
        alt="weedle"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/13.svg"
      />
      <h2
        class="card__title"
      >
        weedle
      </h2>
    </a>
    <a
      class="card"
      href="/kakuna"
    >
      <img
        alt="kakuna"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/14.svg"
      />
      <h2
        class="card__title"
      >
        kakuna
      </h2>
    </a>
    <a
      class="card"
      href="/beedrill"
    >
      <img
        alt="beedrill"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/15.svg"
      />
      <h2
        class="card__title"
      >
        beedrill
      </h2>
    </a>
    <a
      class="card"
      href="/pidgey"
    >
      <img
        alt="pidgey"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/16.svg"
      />
      <h2
        class="card__title"
      >
        pidgey
      </h2>
    </a>
    <a
      class="card"
      href="/pidgeotto"
    >
      <img
        alt="pidgeotto"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/17.svg"
      />
      <h2
        class="card__title"
      >
        pidgeotto
      </h2>
    </a>
    <a
      class="card"
      href="/pidgeot"
    >
      <img
        alt="pidgeot"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/18.svg"
      />
      <h2
        class="card__title"
      >
        pidgeot
      </h2>
    </a>
    <a
      class="card"
      href="/rattata"
    >
      <img
        alt="rattata"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/19.svg"
      />
      <h2
        class="card__title"
      >
        rattata
      </h2>
    </a>
    <a
      class="card"
      href="/raticate"
    >
      <img
        alt="raticate"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/20.svg"
      />
      <h2
        class="card__title"
      >
        raticate
      </h2>
    </a>
    <a
      class="card"
      href="/spearow"
    >
      <img
        alt="spearow"
        class="card__image"
        loading="lazy"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/21.svg"
      />
      <h2
        class="card__title"
      >
        spearow
      </h2>
    </a>
  </div>
</div>
`;
